### PR TITLE
Add Transceiver Responder Code

### PR DESCRIPTION
**Pull request by nomis52**

### CODE REVIEW

**Comment by nomis52** [comment]
This still requires work. I just wanted to see what broke.

File: firmware/src/transceiver.c [hunk 1]
```diff
@@ -137,86 +191,260 @@ static TransceiverHardwareSettings g_hw_settings;
...
+
+void UART_FlushRX() {
+  while (PLIB_USART_ReceiverDataIsAvailable(g_hw_settings.usart)) {
+    PLIB_USART_ReceiverByteReceive(g_hw_settings.usart);
+  }
 }
 
 /*
- * @brief Put us into a MARK state
+ * @brief Pull data out the UART RX queue.
```
Comment (peternewman) [inline firmware/src/transceiver.c]: SPaG out of

File: firmware/src/transceiver.c [hunk 2]
```diff
@@ -137,86 +191,260 @@ static TransceiverHardwareSettings g_hw_settings;
...
+
+/*
+ * @brief Setup the transceiver buffers.
+ */
+static void InitializeBuffers() {
+  g_transceiver.active = NULL;
+  g_transceiver.next = NULL;
+
+  int i = 0;
+  for (; i < NUMBER_OF_BUFFERS; i++) {
```
Comment (peternewman) [inline firmware/src/transceiver.c]: Any reason the i isn't in the for loop definition?

File: firmware/src/transceiver.c [hunk 3]
```diff
@@ -297,47 +523,98 @@ void __ISR(_INPUT_CAPTURE_2_VECTOR, ipl6) InputCaptureEvent(void) {
...
+      case STATE_R_RX_PREPARE:
+      case STATE_R_TX_WAITING:
+      case STATE_R_TX_BREAK:
+      case STATE_R_TX_MARK:
+      case STATE_R_TX_DATA:
+      case STATE_R_TX_DRAIN:
+      case STATE_R_TX_COMPLETE:
+      case STATE_ERROR:
       case STATE_RESET:
         // Should never happen.
```
Comment (peternewman) [inline firmware/src/transceiver.c]: Log a debug message if it does?

File: firmware/src/transceiver.c [hunk 4]
```diff
@@ -799,40 +1126,151 @@ void Transceiver_Tasks() {
...
+      PLIB_IC_Disable(INPUT_CAPTURE_MODULE);
+      PLIB_IC_FirstCaptureEdgeSelect(INPUT_CAPTURE_MODULE, IC_EDGE_FALLING);
+      PLIB_IC_Enable(INPUT_CAPTURE_MODULE);
+      SYS_INT_SourceEnable(INT_SOURCE_INPUT_CAPTURE_2);
+
+      // Fall through
+    case STATE_R_RX_MBB:
+      // noop, waiting for IC event
+
+      /*
```
Comment (peternewman) [inline firmware/src/transceiver.c]: Remove this?

**Comment by peternewman** [comment]
LGTM, some comments.

File: .travis.yml [hunk 5]
```diff
@@ -77,7 +77,7 @@ before_install:
 
 after_success:
  - if [ "$TASK" = "coverity" ]; then cat /home/travis/build/${TRAVIS_REPO_SLUG}/tests/cov-int/scm_log.txt; fi
- - if [ "$TASK" = "coverage" ]; then cd tests && coveralls --gcov /usr/bin/gcov-5 -b . -e include -e mocks -e config.h -e system_config -E '.*Test\.cpp$' --gcov-options '\-lp' > /dev/null; fi
+ - if [ "$TASK" = "coverage" ]; then cd tests && coveralls --gcov /usr/bin/gcov-5 -b . -e include -e harmony -e mocks -e config.h -e system_config -E '.*Test\.cpp$' --gcov-options '\-lp' > /dev/null; fi
```
Comment (nomis52) [inline .travis.yml]: I'm a bit confused as to why we have nothing in the coverage data now.

File: firmware/src/transceiver.c (hunk 1 above)
Comment (nomis52) [inline firmware/src/transceiver.c]: Fixed

File: firmware/src/transceiver.c (hunk 2 above)
Comment (nomis52) [inline firmware/src/transceiver.c]: MPLAB uses ANSI C style so it complains.

File: firmware/src/transceiver.c (hunk 3 above)
Comment (nomis52) [inline firmware/src/transceiver.c]: This is in an interrupt handler so we can't log.

File: firmware/src/transceiver.c (hunk 4 above)
Comment (nomis52) [inline firmware/src/transceiver.c]: I still need to figure out how to handle this.

File: firmware/src/transceiver.c (hunk 2 above)
Comment (peternewman) [inline firmware/src/transceiver.c]: Ah okay.

File: firmware/src/transceiver.c (hunk 3 above)
Comment (peternewman) [inline firmware/src/transceiver.c]: Okay.

File: firmware/src/transceiver.c (hunk 4 above)
Comment (peternewman) [inline firmware/src/transceiver.c]: TODO then?

File: .travis.yml (hunk 5 above)
Comment (peternewman) [inline .travis.yml]: It looks like they've release the fix, so it would be good to get a commit in and test it:
https://pypi.python.org/pypi/cpp-coveralls/0.3.8

Comment (nomis52) [inline .travis.yml]: Done in a separate PR. https://github.com/OpenLightingProject/ja-rule/pull/63

**Comment by peternewman** [comment]
Is this ready for review again now?

**Comment by nomis52** [comment]
Yeah, let's go ahead and get this in. There is still some to do but I'll do that in a separate PR since this one is getting pretty big.

File: firmware/src/constants.h [hunk 6]
```diff
@@ -239,23 +239,24 @@ typedef enum {
...
+  RC_OK = 0,  //!< The command completed successfully.
+  RC_UNKNOWN = 1,  //!< Unknown command
   /**
    * @brief The command could not be completed due to a full memory buffer
    */
-  RC_BUFFER_FULL,
-  RC_BAD_PARAM,  //!< The command was malformed.
-  RC_TX_ERROR,  //!< There was an error during transceiver transmit.
-  RC_RDM_TIMEOUT,  //!< No response was received.
+  RC_BUFFER_FULL = 2,
```
Comment (peternewman) [inline firmware/src/constants.h]: Doxygen

File: firmware/src/constants.h [hunk 7]
```diff
@@ -239,23 +239,24 @@ typedef enum {
...
+  RC_BAD_PARAM = 3,  //!< The command was malformed.
+  RC_TX_ERROR = 4,  //!< There was an error during transceiver transmit.
+  RC_RDM_TIMEOUT = 5,  //!< No response was received.
```
Comment (peternewman) [inline firmware/src/constants.h]: No RDM response...

File: firmware/src/constants.h [hunk 8]
```diff
@@ -239,23 +239,24 @@ typedef enum {
...
 
   /**
    * @brief Data was received in response to a broadcast RDM command.
    *
    * This usually indicates a broken responder.
    */
-  RC_RDM_BCAST_RESPONSE,
-  RC_RDM_INVALID_RESPONSE,  //!< An invalid RDM response was received.
+  RC_RDM_BCAST_RESPONSE = 6,
```
Comment (peternewman) [inline firmware/src/constants.h]: Doxygen

File: firmware/src/constants.h (hunk 6 above)
Comment (peternewman) [inline firmware/src/constants.h]: Ignore me.

File: firmware/src/constants.h (hunk 8 above)
Comment (peternewman) [inline firmware/src/constants.h]: And again.

File: firmware/src/system_config/default/system_pipeline.h [hunk 9]
```diff
@@ -48,16 +48,26 @@ extern "C" {
...
 
 /**
+ * @brief Specifies the function to call when a transceiver RX event occurs.
+ * @sa TransceiverEventCallback.
+ *
+ * This should either call a function of type TransceiverEventCallback or
+ * undefined.
+ */
+#define PIPELINE_TRANSCEIVER_RX_EVENT(event) \
+  Responder_Receive(event);
```
Comment (peternewman) [inline firmware/src/system_config/default/system_pipeline.h]: Indent to four.

File: firmware/src/system_config/default/system_pipeline.h [hunk 10]
```diff
@@ -65,7 +75,8 @@ extern "C" {
  * This should either call a function of type RDMResponderSendCallback or
  * undefined.
  */
-//#define PIPELINE_RDMRESPONDER_SEND(include_break, iov, iov_len) \
+#define PIPELINE_RDMRESPONDER_SEND(include_break, iov, iov_len) \
+  Transceiver_QueueRDMResponse(include_break, iov, iov_len);
```
Comment (peternewman) [inline firmware/src/system_config/default/system_pipeline.h]: Again.

File: firmware/src/system_config/mx_795_512l/system_pipeline.h [hunk 11]
```diff
@@ -22,9 +22,15 @@ extern "C" {
 #define PIPELINE_LOG_WRITE(message) \
   USBConsole_Log(message);
 
-#define PIPELINE_TRANSCEIVER_EVENT(event) \
+#define PIPELINE_TRANSCEIVER_TX_EVENT(event) \
   MessageHandler_TransceiverEvent(event);
```
Comment (peternewman) [inline firmware/src/system_config/mx_795_512l/system_pipeline.h]: Again

File: firmware/src/system_config/mx_795_512l/system_pipeline.h [hunk 12]
```diff
@@ -22,9 +22,15 @@ extern "C" {
...
 
+#define PIPELINE_TRANSCEIVER_RX_EVENT(event) \
+  Responder_Receive(event);
```
Comment (peternewman) [inline firmware/src/system_config/mx_795_512l/system_pipeline.h]: Again

File: firmware/src/system_config/mx_795_512l/system_pipeline.h [hunk 13]
```diff
@@ -22,9 +22,15 @@ extern "C" {
...
+
+#define PIPELINE_RDMRESPONDER_SEND(include_break, iov, iov_len) \
+  Transceiver_QueueRDMResponse(include_break, iov, iov_len);
```
Comment (peternewman) [inline firmware/src/system_config/mx_795_512l/system_pipeline.h]: Again

File: firmware/src/transceiver.c [hunk 14]
```diff
@@ -43,34 +43,73 @@
 
 #define MARK_FUDGE_FACTOR 217
 
+// TODO(simon): retime this.
```
Comment (peternewman) [inline firmware/src/transceiver.c]: Does this need doing before merging?

File: firmware/src/transceiver.c [hunk 15]
```diff
@@ -43,34 +43,73 @@
...
+  STATE_R_RX_MBB = 16,  //!< In mark before break
+  STATE_R_RX_BREAK = 17,  //!< In break
+  STATE_R_RX_MARK = 18,  //!< In mark after break
+  STATE_R_RX_DATA = 19,  //!< Receiving data
+  STATE_R_TX_WAITING = 20,  //!< Delay before response
+  STATE_R_TX_BREAK = 21,  //!< In TX Break
+  STATE_R_TX_MARK = 22,  //!< In TX Mark
+  STATE_R_TX_DATA = 23,  //!< Transmitting data.
+  STATE_R_TX_DRAIN = 24,  //!< Wait for last byte to be sent.
+  STATE_R_TX_COMPLETE = 80,
```
Comment (peternewman) [inline firmware/src/transceiver.c]: Why the jump?

File: firmware/src/transceiver.c [hunk 16]
```diff
@@ -140,86 +203,233 @@ static TransceiverHardwareSettings g_hw_settings;
...
+  PLIB_USART_ReceiverDisable(g_hw_settings.usart);
+  PLIB_USART_TransmitterInterruptModeSelect(g_hw_settings.usart,
+                                            USART_TRANSMIT_FIFO_EMPTY);
+
+  TakeNextBuffer();
+
+  // Enable timer
+  // TODO(simon): Use rdm_responder_delay & rdm_responder_jitter to make this
+  // configurable.
+  PLIB_TMR_Period16BitSet(TMR_ID_3, 1760 - RESPONSE_FUDGE_FACTOR);
```
Comment (peternewman) [inline firmware/src/transceiver.c]: Where has 1760 come from?

File: firmware/src/transceiver.c [hunk 17]
```diff
@@ -227,65 +437,61 @@ static inline void Transceiver_LogStateChange() {
...
- */
-void Transceiver_InitializeBuffers() {
-  g_transceiver.active = NULL;
-  g_transceiver.next = NULL;
+static inline void RXFrameEvent() {
+  TransceiverEvent event = {
+    0,
+    T_OP_RX,
+    g_transceiver.event_index == 0 ? T_RESULT_RX_START_FRAME :
+      T_RESULT_RX_CONTINUE_FRAME,
```
Comment (peternewman) [inline firmware/src/transceiver.c]: Indent four from g_ or to line up with T_RESULT

File: firmware/src/transceiver.c [hunk 18]
```diff
@@ -302,47 +510,98 @@ void __ISR(_INPUT_CAPTURE_2_VECTOR, ipl6) InputCaptureEvent(void) {
...
+          g_timing.request.mark_time = value - g_timing.request.break_time;
+          g_transceiver.state = STATE_R_RX_DATA;
+        }
+        break;
+
+      case STATE_R_RX_DATA:
+        g_transceiver.last_change = value;
+        break;
+
+      case STATE_C_INITIALIZE:
```
Comment (peternewman) [inline firmware/src/transceiver.c]: Fall through comments to keep coverity happy.

File: firmware/src/transceiver_timing.h [hunk 19]
```diff
@@ -24,6 +24,11 @@
  * @addtogroup transceiver
  * @{
  * @file transceiver_timing.h
+ * @brief The timing constants for the 485 Transceiver.
+ *
+ * This header contains the timing constants from E1.11-2008 [DMX] and
+ * E1.20-2010 [RDM]. In some cases we allow the user to configure values
+ * outside the range in t he standards. This is for testing purposes.
```
Comment (peternewman) [inline firmware/src/transceiver_timing.h]: SPaG the

Comment (peternewman) [inline firmware/src/transceiver_timing.h]: SPaG standards; this

**Comment by peternewman** [comment]
LGTM, comments are mostly minor.

File: firmware/src/constants.h (hunk 7 above)
Comment (nomis52) [inline firmware/src/constants.h]: Done

File: firmware/src/system_config/default/system_pipeline.h (hunk 9 above)
Comment (nomis52) [inline firmware/src/system_config/default/system_pipeline.h]: These are like functions, so they are 2 space indents.

File: firmware/src/transceiver.c (hunk 14 above)
Comment (nomis52) [inline firmware/src/transceiver.c]: I'll do this after I finished the rest of the cleanup.

File: firmware/src/transceiver.c (hunk 15 above)
Comment (nomis52) [inline firmware/src/transceiver.c]: I didn't know how many I'd need, fixed.

File: firmware/src/transceiver.c (hunk 16 above)
Comment (nomis52) [inline firmware/src/transceiver.c]: 176us is the min response time. The next change will make this configurable.

File: firmware/src/transceiver.c (hunk 17 above)
Comment (nomis52) [inline firmware/src/transceiver.c]: Done

File: firmware/src/system_config/default/system_pipeline.h (hunk 9 above)
Comment (peternewman) [inline firmware/src/system_config/default/system_pipeline.h]: Okay

File: firmware/src/transceiver.c (hunk 14 above)
Comment (peternewman) [inline firmware/src/transceiver.c]: Okay

File: firmware/src/transceiver.c (hunk 16 above)
Comment (peternewman) [inline firmware/src/transceiver.c]: Right

File: firmware/src/transceiver.c (hunk 18 above)
Comment (peternewman) [inline firmware/src/transceiver.c]: Actually this may only be needed if there is code inbetween the cases.

**Comment by peternewman** [comment]
Can you fix the last few comments, then we're good to go.

File: firmware/src/transceiver.c (hunk 18 above)
Comment (nomis52) [inline firmware/src/transceiver.c]: Correct.

File: firmware/src/transceiver_timing.h (hunk 19 above)
Comment (nomis52) [inline firmware/src/transceiver_timing.h]: Done

File: firmware/src/transceiver.c (hunk 18 above)
Comment (nomis52) [inline firmware/src/transceiver.c]: Correct

**Comment by peternewman** [comment]
LGTM
